File: javascript/124.binary-tree-maximum-path-sum.js
/*
 * @lc app=leetcode id=124 lang=javascript
 *
 * [124] Binary Tree Maximum Path Sum
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var maxPathSum = function(root) {
  // NOTE: Revisit when wiser
  // O(n) time
  // O(log(n)) space
  let maxPathVal = root.val
  const dfs = (root) => {
    if (!root) return 0

    let leftMax = Math.max(dfs(root.left), 0)
    let rightMax = Math.max(dfs(root.right), 0)

    // compute max path sum WITH split
    maxPathVal = Math.max(maxPathVal, root.val + leftMax + rightMax)

    return root.val + Math.max(leftMax, rightMax)
  }

  dfs(root)
  return maxPathVal
};
// @lc code=end

File: cpp/124.binary-tree-maximum-path-sum.cpp
/*
 * @lc app=leetcode id=124 lang=cpp
 *
 * [124] Binary Tree Maximum Path Sum
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution
{
public:
    int maxPathVal;
    int dfsHelper(TreeNode *root)
    {
        if (!root)
            return 0;
        int leftMax = max(dfsHelper(root->left), 0);
        int rightMax = max(dfsHelper(root->right), 0);

        maxPathVal = max(maxPathVal, root->val + leftMax + rightMax);
        return root->val + max(leftMax, rightMax);
    }
    int maxPathSum(TreeNode *root)
    {
        // calculate max of left and right at every node
        if (!root)
            return 0;
        maxPathVal = root->val;
        dfsHelper(root);
        return maxPathVal;
    }
};
// @lc code=end
